kv/memberlist/http_status_handler.go

```diff
@@ -0,0 +1,217 @@
+type StatusPageData struct {
+	Now              time.Time
+	Memberlist       *memberlist.Memberlist
+	SortedMembers    []*memberlist.Node
+	Store            map[string]ValueDesc
+	SentMessages     []Message
+	ReceivedMessages []Message
+}
+
+// NewHTTPStatusHandler creates a new HTTPStatusHandler that will render the provided template using the data from StatusPageData.
+func NewHTTPStatusHandler(kvs *KVInitService, tpl *template.Template) HTTPStatusHandler {
+	return HTTPStatusHandler{kvs, tpl}
+}
+
+func (h HTTPStatusHandler) ServeHTTP(w http.ResponseWriter, req *http.Request) {
+	kv := h.kvs.getKV()
+	if kv == nil {
+		w.Header().Set("Content-Type", "text/plain")
+		// Ignore inactionable errors.
+		_, _ = w.Write([]byte("This instance doesn't use memberlist."))
+		return
+	}
+
+	const (
+		downloadKeyParam    = "downloadKey"
+		viewKeyParam        = "viewKey"
+		viewMsgParam        = "viewMsg"
+		deleteMessagesParam = "deleteMessages"
+	)
+
+	if err := req.ParseForm(); err == nil {
+		if req.Form[downloadKeyParam] != nil {
+			downloadKey(w, kv, kv.storeCopy(), req.Form[downloadKeyParam][0]) // Use first value, ignore the rest.
+			return
+		}
+
+		if req.Form[viewKeyParam] != nil {
+			viewKey(w, kv.storeCopy(), req.Form[viewKeyParam][0], getFormat(req))
+			return
+		}
+
+		if req.Form[viewMsgParam] != nil {
+			msgID, err := strconv.Atoi(req.Form[viewMsgParam][0])
+			if err != nil {
+				http.Error(w, err.Error(), http.StatusBadRequest)
+				return
+			}
+
+			sent, received := kv.getSentAndReceivedMessages()
+
+			for _, m := range append(sent, received...) {
+				if m.ID == msgID {
+					viewMessage(w, kv, m, getFormat(req))
+					return
+				}
+			}
+
+			http.Error(w, "message not found", http.StatusNotFound)
+			return
+		}
+
+		if len(req.Form[deleteMessagesParam]) > 0 && req.Form[deleteMessagesParam][0] == "true" {
+			kv.deleteSentReceivedMessages()
+
+			// Redirect back.
+			w.Header().Set("Location", "?"+deleteMessagesParam+"=false")
+			w.WriteHeader(http.StatusFound)
+			return
+		}
+	}
+
+	members := kv.memberlist.Members()
+	sort.Slice(members, func(i, j int) bool {
+		return members[i].Name < members[j].Name
+	})
+
+	sent, received := kv.getSentAndReceivedMessages()
+
+	v := StatusPageData{
+		Now:              time.Now(),
+		Memberlist:       kv.memberlist,
+		SortedMembers:    members,
+		Store:            kv.storeCopy(),
+		SentMessages:     sent,
+		ReceivedMessages: received,
+	}
+
+	accept := req.Header.Get("Accept")
+	if strings.Contains(accept, "application/json") {
+		w.Header().Set("Content-Type", "application/json")
+
+		if err := json.NewEncoder(w).Encode(v); err != nil {
+			http.Error(w, err.Error(), http.StatusInternalServerError)
+		}
+		return
+	}
+
+	w.Header().Set("Content-Type", "text/html")
+	if err := h.tpl.Execute(w, v); err != nil {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+	}
+}
+
+func getFormat(req *http.Request) string {
```

Comment:
Note to reviewers: these utility functions have just been copied as-is from `kv_init_service.go`